[global] Update config file location and layout

TurboTurtle opened the pull request:
Moves the default config file we look for to /etc/sos/sos.conf instead
of /etc/sos.conf. Extends the config file to look for a section matching
the name of the component being used. Renames the "general" section to
"global" and the "tunables" section to "plugin_options".

Updates the default sos.conf to this style and adds some comments to the
file. Update the man page for sos.conf.

Note that this commit does NOT update sos.spec to drop the default
configuration file in the new location, as that will be handled by a
later commit to update the specfile wholesale.

Closes: #2125
Resolves: #2136

Signed-off-by: Jake Hunsaker <jhunsake@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] If this commit closes an existing issue, is the line `Closes: #ISSUENUMBER` included in an independent line?
- [x] If this commit resolves an existing pull request, is the line `Resolves: #PRNUMBER` included in an independent line?

TurboTurtle commented:
Note: this commit does *not* implement the per-user overrides that I briefly mentioned in #2125. If this design is acceptable, then I can either extend this PR or we can add that later as we close out 4.0 (or go into 4.1 to ensure stability of everything else going into 4.0).

pmoravec commented:
I think no option from `[global]` section in `/etc/sos/sos.conf` have an effect, currently. E.g. having there:

    [global]
    batch = true

then the `batch` option is not applied to either `report` or `clean`. The same with `verbose = 3` - default verbosity used instead.

TurboTurtle commented:
Oops, left off a conditional check there for global. Fixed now.

pmoravec commented:
./sos/__init__.py:                                dest="config_file", default="/etc/sos.conf",


needs to change `default` to `/etc/sos/sos.conf` :)

pmoravec commented:
The parser ignores unknown options - we should at least warn about them (cf. `sos report --blah` returns an error, even).

Adding an option twice to the same section, like:

    [report]
    build = true
    build = true

then `sos` fails with:

    sosreport (version 3.9)

    Failed to parse configuration file /etc/sos/sos.conf

and even verbose logging does not suggest why.


Neither issue is a regression, the same behaviour applies to the current code, so I can raise a separate issue for it (until @TurboTurtle decides to fix it within this PR).

pmoravec commented on sos.conf:
```diff
@@ -1,11 +1,36 @@
-#disable = rpm, selinux, dovecot
+[report]
+# Options that will apply to any `sos report` run should be listed here.
+# Note that the option names *must* be the long-form name as seen in --help
```
There is something wrong.. 2 issues in fact:

- `sos report has cmd option `--skip-plugins` but we support `noplugins` option in config file.
- `sos report has cmd option `--only-plugins` but we support `onlyplugins` option in config file - without the "`-`". Config file option `only-plugins = kernel` is ignored.

TurboTurtle replied:
Nice catch, we'll need to change the destination in the `add_argument()` statement for these, or change the long form syntax.

TurboTurtle replied:
Rather, the above is just for `-n`. We use `skip-plugins` in the long form, but `noplugins` in code. I think `skip-plugins` would be more favorable here, and I'll go with that if no one objects.

TurboTurtle commented:
> The parser ignores unknown options - we should at least warn about them (cf. `sos report --blah` returns an error, even).
> Adding an option twice to the same section, like:
> Neither issue is a regression, the same behaviour applies to the current code, so I can raise a separate issue for it (until @TurboTurtle decides to fix it within this PR).

Let's fix it here, all at once. Notifying of invalid options is no issue, however I'm not sure why duplicate options in the same section would cause that issue, unless it's built into `ConfigParser()` to fail like that.

TurboTurtle commented:
Updated the PR with:

* Update the arg default for config file within the report parser as well
* Catch `DuplicateOptionError` separately and report it (and abort, `ConfigParser()` does not offer a way to recover gracefully)
* Catch unknown args and report them, but do not abort
* After config file parsing is done, log the updated effective options
* Standardize the option long-form names and the destination vars they save to, e.g. `noplugins` -> `skip_plugins`.

TurboTurtle commented:
I'm going to work towards extending this PR to include the per-user config file overrides, it just makes sense to do it all at once here.

Are we good with the proposal in #2125 to move root presets/host groups into `/etc/sos/$thing.d/` and non-root to `~/.config/sos/$thing.d/`, or should we place them elsewhere? Or leave everything alone right now and have it still live in `/var/lib/sos/`?

TurboTurtle commented:
Added a few more commits to this series, that independently update the `report` and `collect` components to use `/etc/sos` now instead of `/var/lib/sos|sos-collector`.

bmr-cymru commented:
Since this will affect packagers of `sos` I think we ought to mention it prominently in the release notes - I thought we already had a label for that, but I couldn't find one so I've created the `release note` label and added it to this PR.

TurboTurtle commented:
Noticed that I had inadvertently swapped the precedence order of cmdline vs config file. That's now resolved, and with it I've moved the user config file handling into `SoSComponent.load_options()` rather than in `_init_component()`.

However, I was not able to do this using the `merge()` function, due to how argparse passes parsed cmdline arguments, which was causing `merge()` to completely override the config file (effectively negating it). So, I've found the need to manually parse the non-default options out from the cmdline option object, and use that to update the `SoSOptions` instantiated and updated from the config files.

pmoravec commented:
> Implementation looks good. A bit tangential to this, but since we now have config file(s), presets, command line, I wonder if we should clarify in the docs the precedence for options? If I remember rightly it's conf < preset < command line?

+1 to make this documented. IMHO the conf < preset < cmdline is right (or maybe even preset < conf < cmdline, since preset is "product default", while conf is "system default", i.e. more specific settings?).

Current behaviour of the precedence is bit blurred to me:
1) having `/etc/sos/sos.conf`:

    [global]
    verbose = 3
    log-size = 1

    [report]
    log-size = 2

Then the `log-size` option is fully ignored (either I write the option to either section). `verbose = 3` is taken into account from either location (which is good), just the `log-size` from neither.

2) Mimicking I am on Satellite (by installing a dummy "satellite-common" package), then `sosreport --log-size=3` loads and uses `log-size=100` from the presets - i.e. the real precedence is cmdline < presets.

TurboTurtle commented:
> +1 to make this documented. IMHO the conf < preset < cmdline is right (or maybe even preset < conf < cmdline, since preset is "product default", while conf is "system default", i.e. more specific settings?).
> 

Done. This is now at the top of `man sos.conf`.
 
> Then the `log-size` option is fully ignored (either I write the option to either section). `verbose = 3` is taken into account from either location (which is good), just the `log-size` from neither.

Fixed. Had a typo that was causing the `arg_defaults` to be set in correctly and thus log-size was being ignored.
> 
> 1. Mimicking I am on Satellite (by installing a dummy "satellite-common" package), then `sosreport --log-size=3` loads and uses `log-size=100` from the presets - i.e. the real precedence is cmdline < presets.

Fixed. We now re-apply cmdline overrides after presets are loaded.

pmoravec commented on sos/__init__.py:
```diff
@@ -20,6 +20,7 @@
 import sys
 
 from argparse import ArgumentParser
+from pathlib import Path
```
I think this is not further required import (while/though I am curious why Travis didnt detect it as superfluous).

pmoravec commented on sos/component.py:
```diff
@@ -169,12 +204,13 @@ def load_options(self):
-        cmdopts = SoSOptions().from_args(self.parser.parse_args(self.cmdline))
-        opts.merge(cmdopts)
+        opts.update_from_conf(self.args.config_file, self.args.component)
+        if os.getuid() != 0:
```
Why `root` cant use its own config file as well? I am quite OK with this, just dont see much rationale (can't root require different defaults than system ones?).

TurboTurtle replied:
I don't mind removing the restriction, but my rationale here was simply that it's not a common trend.

pmoravec commented on sos/options.py:
```diff
@@ -237,7 +239,7 @@ def merge(self, src, skip_default=True):
                 self._merge_opt(arg, src, False)
 
-    def dict(self):
+    def dict(self, preset_filter=True):
```
A nice catch to still apply `--preset` cmdline option.

pmoravec commented on man/en/sos.conf.5:
```diff
@@ -51,19 +119,23 @@ threads=10
+[report]
 .br
-disable = host, filesys
+noplugins = host,filesys
```
This is renamed to `skip-plugins` else-where.

I see this renaming bit controversial, as 1) it is backward incompatible, 2) short option `-n` is not a shortcut from `--skip-plugins` (while we already use `-s` for `--sysroot`). So (not ultimate) +1 for keeping the old "noplugins" or could-be "no-plugins" option.

TurboTurtle replied:
Hmm. I don't think we'll be able to avoid point #1, but as this is a major version update I'm also not too concerned about that. 

The problem I see is that both the short- and long-form name make sense in their own context. `-n` as a "do *n*ot use this" is fairly conventional. `skip-plugins` is also straight forward and intuitive.

However, a long form of `no-plugins` doesn't sound right to an English speaker. I wouldn't say "run sos with no plugins kvm, host, logs", and the first thing that comes to mind with `no-plugins` is "oh, this option means disable all plugins" which obviously isn't the case.

pmoravec commented on sos/collector/__init__.py:
```diff
@@ -32,7 +33,7 @@
 from sos import __version__
 
-COLLECTOR_LIB_DIR = '/var/lib/sos-collector'
+COLLECTOR_LIB_DIR = '/etc/sos/groups.d'
```
`s/COLLECTOR_LIB_DIR/COLLECTOR_CONFIG_DIR/` or similar?

pmoravec commented on sos/collector/__init__.py:
```diff
@@ -607,12 +614,16 @@ def write_host_group(self):
-        if not os.path.isdir(COLLECTOR_LIB_DIR):
-            raise OSError("%s no such directory" % COLLECTOR_LIB_DIR)
-        fname = COLLECTOR_LIB_DIR + '/' + cfg['name']
+        if os.getuid() != 0:
```
Ditto on another place why `root` cant have its own `.config` (but again, I am ok with the current behaviour).

pmoravec commented on man/en/sos.conf.5:
```diff
@@ -3,7 +3,75 @@
+  2. User-specific configuration file at $HOME/.config/sos/sos.conf (for sos
+     components that support non-root)
+  3. In the case of running \fBsos report\fR, presets either automatically loaded
+     due to system configuration, or specified via \fB--add-preset\fR
```
.. specified via `--add-preset` or just `--preset` (also)? The later worked to me as well.

TurboTurtle replied:
Whoops, that's a typo.

pmoravec commented:
Sorry for the iterative comments with such delays i between.. And thanks for great PR in general.

Few relics point to old config file location:

- https://github.com/sosreport/sos/blob/master/tests/simple.sh#L85-L91
- https://github.com/sosreport/sos/blob/master/sos/report/plugins/host.py#L33 (worth copying both old and new config file for legacy reasons, or is it obsolete and the new one is enough?)


ACK otherwise!

TurboTurtle commented:
Forgot to mention. This last push:

- Updates the `host` plugin, and adds a forbidden path on `default_mapping`
- Updates `simple.sh`
- Fixes the man page typo, and also cleans up the formatting a bit for the `extras.d/` section
- Changes `COLLECTOR_LIB_DIR` to `COLLECTOR_CONFIG_DIR`
- Updates `noplugins` to `skip-plugins` in the default `sos.conf`. I've left it as `skip-plugins` for the reasons I gave above, but I'm certainly open for counter arguments.

pmoravec commented on man/en/sos.conf.5:
```diff
@@ -51,19 +120,23 @@ threads=10
-/etc/sos.conf
+/etc/sos/sos.conf
+.br
+$HOME/sos/sos.conf (optional)
```
Rather `$HOME/.config/sos/sos.conf` ?

pmoravec commented on tests/simple.sh:
```diff
@@ -83,11 +83,12 @@ run_expecting_sucess () {
 }
 
 # If /etc/sos.conf doesn't exist let's just make it..
```
The comment lacks "sos" subdir (as well as "Creating .. few lines below").

pmoravec commented on sos/report/plugins/host.py:
```diff
@@ -20,6 +20,8 @@ class Host(Plugin, RedHatPlugin, DebianPlugin):
 
     def setup(self):
 
+        self.add_forbidden_path('/etc/sos/default_mapping')
```
Great catch!

Just.. isn't the path `/etc/sos/cleaner/default_mapping` instead?

TurboTurtle replied:
Yup, I forgot we moved that into the `cleaner` subdir. Updating these shortly.

pmoravec commented:
`sos_extras` plugin should replace `sos.extras.d` by `sos/extras.d` (in `short_desc` and namely in `extras_dir`).

pmoravec commented:
ACK to the PR, appart of the simple / final issues spotted now.
